### PR TITLE
Change logfile names and also allow plasma store socket to be passed in.

### PR DESCRIPTION
**Pull request by suquark**
## What do these changes do?

This PR resolves Ray's temporary file/socket creating problems.

Before this PR, most of Ray's temp files and sockets are created with a self-designed random name under fixed `/tmp` directory. This is problematic because:

* It does not check the existing of a file and hopes we are lucky.
* It is not cross-platform friendly. Systems like windows may not have `/tmp`
* Users are unable to control the location of these files. This makes it hard to share the socket with others through Docker volume mounts
* Temp files are hard to track and mixed up under `/tmp`.
* We have no way to reconnect to a previous socket.
* Logging files & socket files have confusing random numbers in their names.

This PR fixes these problems.

Look into `doc/source/tempfile.rst` for details.

## Related issue number

https://github.com/ray-project/ray/issues/2686
https://github.com/ray-project/ray/issues/2687
https://github.com/ray-project/ray/issues/2594 (partly)

### CODE REVIEW

File: python/ray/tempfile_services.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+
+def get_random_ipython_notebook_path(port):
+    """Get a new random ipython notebook path"""
+
+    notebook_filepath = os.path.join(
+        os.path.dirname(os.path.abspath(__file__)), "WebUI.ipynb")
+    # We copy the notebook file so that the original doesn't get modified by
+    # the user.
+    notebook_name = tempfile.mktemp(suffix='.ipynb', prefix='ray_ui_',
```
Comment (mitar) [inline python/ray/tempfile_services.py]: `mktemp` is deprecated.

File: python/ray/tempfile_services.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        os.chmod(directory_path, 0o0777)
+
+
+if RAY_TEMPDIR in os.environ:
+    temp_root = os.environ[RAY_TEMPDIR]
+    try_to_create_directory(temp_root)
+else:
+    # Change the default root by providing the directory named by the
+    # TMPDIR environment variable.
+    temp_dir = tempfile.TemporaryDirectory(
```
Comment (mitar) [inline python/ray/tempfile_services.py]: You can just use `mkdtemp` here.

File: python/ray/tempfile_services.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+import os
+import shutil
+import tempfile
+
+import ray.utils
+
+logger = logging.getLogger(__name__)
+
+RAY_TEMPDIR = 'RAY_TEMPDIR'
+RAY_OBJECT_STORE_SOCKET_NAME = 'RAY_OBJECT_STORE_SOCKET_NAME'
```
Comment (mitar) [inline python/ray/tempfile_services.py]: Could we have command line arguments configuring this, not environment variables?

**Comment by mitar** [comment]
Thanks for this!

For plasma store socket file, I would prefer if it could be outside of temp directory altogether. So for me it is not enough just to control its filename, but also full location. In this way I can move that to a separate Docker volume.

**Comment by mitar** [comment]
So how I think it should be is that there is a command line to configure the temp location, and if it is not set, it uses Python's default tempfile location for other locations (and default for plasma also). And then if plasma has a command line argument, then it takes precedence.

File: python/ray/tempfile_services.py (hunk 1 above)
Comment (mitar) [reply python/ray/tempfile_services.py]: I think the question is the approach of generating a filename is a good one. I think it is not because it cannot assure unique names. So I would argue that Ray should be changed so that it never uses temporary filename, but uses file objects directly. Could fix #2594.

**Comment by raulchen** [comment]
Thanks for working on this. 

Could you please summarize: 
(1) how do the users specify the location after this change? 
(2) how are the files organized?

I'll take a look at the code today or tomorrow.

**Comment by suquark** [comment]
I have updated the description. This PR is still WIP to fix several incompatibility in tests cause by the changes.

File: python/ray/ray_constants.py [hunk 4]
```diff
@@ -15,6 +15,8 @@ def env_integer(key, default):
 
 
 ID_SIZE = 20
+from ray.local_scheduler import ObjectID
+ray.ObjectID = ObjectID
```
Comment (robertnishihara) [inline python/ray/ray_constants.py]: Why are these two lines this needed?

File: python/ray/scripts/scripts.py [hunk 5]
```diff
@@ -177,11 +177,18 @@ def cli(logging_level, logging_format):
     is_flag=True,
     default=False,
     help="do not redirect non-worker stdout and stderr to files")
+@click.option(
+    "--plasma-store-socket-name",
+    is_flag=True,
+    default=False,
+    help="manually specify the name of the socket of the plasma store"
+)
```
Comment (robertnishihara) [inline python/ray/scripts/scripts.py]: The point of exposing this is so that you can reuse an existing plasma store? Seems reasonable, but do you have a use case in mind?

File: python/ray/tempfile_services.py (hunk 3 above)
Comment (robertnishihara) [reply python/ray/tempfile_services.py]: Let's not use environment variables. If these are needed, please put them in `ray_constants.py`

File: python/ray/tempfile_services.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+import logging
+import os
+import shutil
+import tempfile
+
+import ray.utils
+
+logger = logging.getLogger(__name__)
+
+RAY_TEMPDIR = 'RAY_TEMPDIR'
```
Comment (robertnishihara) [inline python/ray/tempfile_services.py]: In general, use `"` instead of `'`

File: python/ray/tempfile_services.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+            for redirecting stderr. If redirect_output is false, this will
+            return a tuple of two None objects.
+    """
+    if not redirect_output:
+        return None, None
+
+    # Create a directory to be used for process log files.
+    logs_dir = get_logs_dir_path()
+    # Create another directory that will be used by some of the RL algorithms.
+    # TODO: Move it into where we need such a directory.
```
Comment (robertnishihara) [inline python/ray/tempfile_services.py]: When you add `TODO` items, please add your name, (e.g., I use `TODO(rkn)`) so that we know who to ask if we have questions about it.

File: python/ray/tempfile_services.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+
+logger = logging.getLogger(__name__)
+
+RAY_TEMPDIR = 'RAY_TEMPDIR'
+RAY_OBJECT_STORE_SOCKET_NAME = 'RAY_OBJECT_STORE_SOCKET_NAME'
+
+_incremental_dict = collections.defaultdict(lambda :0)
+
+
+def make_inc_temp(suffix="", prefix="", dir=None):
```
Comment (robertnishihara) [inline python/ray/tempfile_services.py]: `dir` is a Python function, so let's use a different name, e.g., `directory_name`

File: python/ray/tempfile_services.py [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+        os.chmod(directory_path, 0o0777)
+
+
+if RAY_TEMPDIR in os.environ:
+    temp_root = os.environ[RAY_TEMPDIR]
+    try_to_create_directory(temp_root)
+else:
+    # Change the default root by providing the directory named by the
+    # TMPDIR environment variable.
+    temp_root = tempfile.mkdtemp(prefix='ray_{pid}_'.format(pid=os.getpid()))
```
Comment (robertnishihara) [inline python/ray/tempfile_services.py]: On my macbook, `tempfile.mkdtemp()` returns `'/var/folders/15/54jf68993rd7753c5fms424r0000gn/T/tmpwtgcu22l'`. This seems substantially less usable than `/tmp/raylogs` because no one will be able to find the directory.

File: python/ray/scripts/scripts.py (hunk 5 above)
Comment (mitar) [reply python/ray/scripts/scripts.py]: That's connected with #2687.

For us this is a way to share data objects computed with Ray with UI tools which can then expose data values which were computed without having to go through some other serialization or memory copying.

File: python/ray/tempfile_services.py (hunk 9 above)
Comment (mitar) [reply python/ray/tempfile_services.py]: This is the [algorithm Python uses for temp dir](https://docs.python.org/3/library/tempfile.html#tempfile.gettempdir).

Comment (robertnishihara) [reply python/ray/tempfile_services.py]: It feels useful to me to have a consistent directory so people know where to look even if they lost the stdout from `ray.init`. It'll have to be a little different on windows, but that doesn't seem too bad.

Comment (mitar) [reply python/ray/tempfile_services.py]: So `mkdtemp` and other methods do get `dir` as an argument where you can set this directory. So then I suggest this is a command line argument which is by default `/tmp` but users can change it. But still, that secure `mkdtemp` and other functions are used, and that it is configurable.

**Comment by suquark** [comment]
@robertnishihara Do you think we need a doc for Ray's temp files?

File: python/ray/tempfile_services.py [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+    date_str = datetime.datetime.today().strftime("%Y-%m-%d_%H-%M-%S")
+    prefix = '{}-{}-'.format(name, date_str)
+
+    log_stdout = make_inc_temp(
+        suffix='.out', prefix=prefix, directory_name=logs_dir)
+    log_stderr = make_inc_temp(
+        suffix='.err', prefix=prefix, directory_name=logs_dir)
+    # Line-buffer the output (mode 1)
+    log_stdout_file = open(log_stdout, "a", buffering=1)
+    log_stderr_file = open(log_stderr, "a", buffering=1)
```
Comment (raulchen) [inline python/ray/tempfile_services.py]: Is it safe to not close these files? 
I used to see a lot of error messages in tests like this (python 3 only):
```
/home/travis/.local/lib/python3.6/site-packages/ray-0.4.0-py3.6-linux-x86_64.egg/ray/services.py:1510: ResourceWarning: unclosed file <_io.TextIOWrapper name='/tmp/raylogs/worker_0_0-2018-07-06_08-19-29-04610.out' mode='a' encoding='UTF-8'> "worker_{}_{}".format(i, j), redirect_output)
```
But these files are created for other processes, I'm not sure what is the best way to close them.

**Comment by raulchen** [comment]
I patched this PR and ran a simple script on my macbook. A couple of temp roots were created under `/var/folders/13/zx68qr9x331c5247wbs60y600000gn/T/`:
```
ray_30566_c7jo5kwt
ray_30586_cdp3k2jm
ray_30588_lo6ncg6v
ray_30604_gac3dn35
ray_30605_m1t24c3q
ray_30606_bt689fvn
ray_30607__zc6nlqh
ray_30635_gqt3ic9u
``` 
Among them, only one is non-empty. And under the non-empty one, the file structure is like the following:
```
/var/folders/13/zx68qr9x331c5247wbs60y600000gn/T/ray_30566_c7jo5kwt
├── logs
│   ├── log_monitor-2018-09-17_11-38-07-1.err
│   ├── log_monitor-2018-09-17_11-38-07-1.out
│   ├── monitor-2018-09-17_11-38-07-1.err
│   ├── monitor-2018-09-17_11-38-07-1.out
│   ├── plasma_manager_0-2018-09-17_11-38-07-1.err
│   ├── plasma_manager_0-2018-09-17_11-38-07-1.out
│   ├── plasma_store_0-2018-09-17_11-38-07-1.err
│   ├── plasma_store_0-2018-09-17_11-38-07-1.out
│   ├── redis-0-2018-09-17_11-38-07-1.err
│   ├── redis-0-2018-09-17_11-38-07-1.out
│   ├── redis-2018-09-17_11-38-07-1.err
│   ├── redis-2018-09-17_11-38-07-1.out
│   ├── webui-2018-09-17_11-38-08-1.err
│   └── webui-2018-09-17_11-38-08-1.out
├── ray
├── ray_ui_1.ipynb
└── sockets
    └── plasma_store_1
```

------

Compared with the previous way, this definitely becomes less usable. Because users would have more difficulties finding the root temp dir.

I'd like to propose an alternative way. Please let me know if you think this makes sense or not.
1) We create a root temp dir for each run under `/tmp/ray/` (configuration), with current date time as suffix. E.g., `/tmp/ray/2018-09-17_11-38-08/`
2) there's a link file `/tmp/ray/latest` that always points to the latest root temp dir.
3) under one root temp dir, files are organized in a few sub dirs by their categories. E.g., `logs`, `sockets`, etc. 

With the above setting, if users want to refer to raylet log of the last run, it's always `/tmp/ray/latest/logs/raylet.out`

**Comment by robertnishihara** [comment]
@suquark it probably makes sense to have a doc on logging in general, part of which is about the log files that are produced (could be in a separate PR).

@raulchen I think that proposal makes a lot of sense (although I think the directory name may need a random string appended in case Ray is started twice at the same time).

File: python/ray/tempfile_services.py (hunk 10 above)
Comment (suquark) [reply python/ray/tempfile_services.py]: These messages have been seen in previous tests. This may be because some process just created some log files and get killed in the test.

**Comment by suquark** [comment]
@raulchen Thanks for your reports and suggestions. I have changed the default temp path into `/tmp` and it should be a bug in code. It's still WIP and I will remove the "WIP" label after it has been stable.

**Comment by mitar** [comment]
> although I think the directory name may need a random string appended in case Ray is started twice at the same time

If a good way to generate files/directories is used (tempfile Python module is such) then only one process can exclusively create a name, and the other is forced to pick another name.

I think we should just keep current `/tmp` location, by default I think this is an easy fix and we do not have to modify that. This PR should just be improving how names are generated (in more secure way, using exclusive generation), and also allow configuration of those paths to non-default values.

**Comment by suquark** [comment]
I have finished this PR. Could you make a further review? @robertnishihara @raulchen

File: doc/source/tempfile.rst [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+---------------------------
+
+First we introduce the concept of a session of Ray.
+
+A session contains a set of processes. A session is created by executing
+``ray start`` command or call `ray.init()` in a Python script and ended by
+executiong ``ray stop`` or call `ray.shutdown()`.
+
+For each session, Ray will create a *root temporary directory* to place all its
+temporary files. The path is `/tmp/ray_{ray_pid}@{datetime}` by default.
```
Comment (mitar) [inline doc/source/tempfile.rst]: I am not sure if I would like to see `@` in the directory name. Also, should we maybe have `ray_{datetime}_{pid}`? Then sorting by name would sort also by time?

File: doc/source/tempfile.rst [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+You are allowed to change the *root temporary directory* in one of these ways:
```
Comment (mitar) [inline doc/source/tempfile.rst]: So when root temporary directory is provided, then pid and datetime is not appended?

File: doc/source/tempfile.rst [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+* Specify `plasma_store_socket_name` when call `ray.init()`
+
+
+Notes
+-----
+
+Temporary file policies are defined in ``python/ray/tempfile_services.py``.
+
+Currently, we keep ``/tmp/ray`` as the default directory for temporary data files of RLlib as before.
+It is not very reasonable and could be changed later.
```
Comment (mitar) [inline doc/source/tempfile.rst]: Why it is not reasonable?

File: doc/source/tempfile.rst [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+
+Plasma Object Store Socket
+--------------------------
+
+Plasma object store sockets can be used to share objects with other programs using Apache Arrow.
+
+You are allowed to specify the plasma object store socket in one of these ways:
+
+* Pass ``--plasma-store-socket-name={your socket path}`` to ``ray start``
+* Specify `plasma_store_socket_name` when call `ray.init()`
```
Comment (mitar) [inline doc/source/tempfile.rst]: This still does not allow me to move plasma store socket to a completely different location, no?

File: doc/source/tempfile.rst (hunk 11 above)
Comment (suquark) [reply doc/source/tempfile.rst]: I use `@` because there are too many `_`s & `-`s in the datetime string and it looks not nice for me. `@` also implies 'at' a certain timestamp and is supported in most platforms without escape charactors like `\`.

File: doc/source/tempfile.rst (hunk 14 above)
Comment (suquark) [reply doc/source/tempfile.rst]: It is a full path, so you can move it to anywhere you like. I should clarify it in the doc.

File: doc/source/tempfile.rst (hunk 13 above)
Comment (suquark) [reply doc/source/tempfile.rst]: Currently, `/tmp/ray` is used as a magic string throughout the code (not feel so happy when  seeing it). I suppose there should be more smart ways but not in this PR.

File: doc/source/tempfile.rst (hunk 12 above)
Comment (suquark) [reply doc/source/tempfile.rst]: Yes. It is not appended because users cannot control the path precisely with pid and datetime. By the way, `pid` is used to avoid inconsistent directory creation (older implementation could crash if we start several Ray processes in parallel).

File: doc/source/tempfile.rst (hunk 11 above)
Comment (raulchen) [reply doc/source/tempfile.rst]: `ray_pid` is the pid of which process? In one session, there will be many processes, right? 
Also, not a big deal, but I'd suggest put these temp dirs under one big root, like `/tmp/ray/${session_root}`. This will make these dirs easier to manage.

File: doc/source/tempfile.rst [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+You can also use ``default_worker.py --temp-dir={your temp path}`` to
+start a new worker with given *root temporary directory*.
+
+The *root temporary directory* you specified will be given as it is
+without pids or datetime attached.
+
+Layout of Temporary Files
+-------------------------
+
+A typical layout of temporary files could look like this:
```
Comment (raulchen) [inline doc/source/tempfile.rst]: maybe we should put a raylet example here, since legacy ray is going to be deprecated soon.

File: doc/source/tempfile.rst (hunk 14 above)
Comment (raulchen) [reply doc/source/tempfile.rst]: considering this, maybe `plasma-store-socket-path` is the better term.

Comment (suquark) [reply doc/source/tempfile.rst]: However, `plasma_store_socket_name` is heavily used in the previous code. The `name` here refers to the name of a unix domain socket. It is not necessary to be a filesystem path. See https://stackoverflow.com/questions/29427769/how-to-use-unix-domain-socket-without-creating-a-socket-file
So I am cautious with changing this name.

File: doc/source/tempfile.rst (hunk 15 above)
Comment (suquark) [reply doc/source/tempfile.rst]: I have adopted your advice. But `plasma_store_socket_name` is reserved. See the related comment above.

File: doc/source/tempfile.rst (hunk 14 above)
Comment (mitar) [reply doc/source/tempfile.rst]: I also think path would be better, if one can provide full path.

File: python/ray/ray_constants.py [hunk 16]
```diff
@@ -4,8 +4,10 @@
 """Ray constants used in the Python code."""
 
 import os
-
-import ray
+# Some tests fail because of circular reference of the `ray` module.
```
Comment (robertnishihara) [inline python/ray/ray_constants.py]: add a newline above and remove comment

File: python/ray/tempfile_services.py [hunk 17]
```diff
@@ -0,0 +1,283 @@
...
+
+
+def make_inc_temp(suffix="", prefix="", directory_name="/tmp"):
+    """Return a incremental temporary file name. The file is not created.
+
+    Args:
+        suffix (str): The suffix of the temp file.
+        prefix (str): The prefix of the temp file.
+        directory_name (str) : The base directory of the temp file.
+    """
```
Comment (robertnishihara) [inline python/ray/tempfile_services.py]: document returns

File: python/ray/tempfile_services.py [hunk 18]
```diff
@@ -0,0 +1,283 @@
...
+
+def get_logs_dir_path():
+    """Get a temp dir for logging."""
+    logs_dir = os.path.join(get_temp_root(), "logs")
+    try_to_create_directory(logs_dir)
+    return logs_dir
+
+
+def get_sockets_dir_path():
+    """Get a temp dir for sockets"""
```
Comment (robertnishihara) [inline python/ray/tempfile_services.py]: period after sockets

File: python/ray/tempfile_services.py [hunk 19]
```diff
@@ -0,0 +1,283 @@
...
+    sockets_dir = os.path.join(get_temp_root(), "sockets")
+    try_to_create_directory(sockets_dir)
+    return sockets_dir
+
+
+def get_raylet_socket_name(suffix=''):
```
Comment (robertnishihara) [inline python/ray/tempfile_services.py]: use `""`

File: python/ray/worker.py [hunk 20]
```diff
@@ -1613,6 +1620,10 @@ def _init(address_info=None,
     node_ip_address = address_info.get("node_ip_address")
     redis_address = address_info.get("redis_address")
 
+    # Because the code will be branched soon, we have to create the temp root
+    # here.
+    tempfile_services.set_temp_root(temp_dir)
```
Comment (robertnishihara) [inline python/ray/worker.py]: I think we don't need this

File: python/ray/scripts/scripts.py [hunk 21]
```diff
@@ -177,11 +177,22 @@ def cli(logging_level, logging_format):
     is_flag=True,
     default=False,
     help="do not redirect non-worker stdout and stderr to files")
+@click.option(
+    "--plasma-store-socket-name",
+    is_flag=True,
+    default=False,
```
Comment (robertnishihara) [inline python/ray/scripts/scripts.py]: change default

File: python/ray/scripts/scripts.py [hunk 22]
```diff
@@ -177,11 +177,22 @@ def cli(logging_level, logging_format):
...
+    help="manually specify the name of the socket of the plasma store")
+@click.option(
+    "--temp-dir",
+    is_flag=True,
```
Comment (robertnishihara) [inline python/ray/scripts/scripts.py]: should not be `is_flag`

File: python/ray/scripts/scripts.py [hunk 23]
```diff
@@ -177,11 +177,22 @@ def cli(logging_level, logging_format):
...
+    default=None,
+    help="manually specify the root temporary dir the Ray process")
```
Comment (robertnishihara) [inline python/ray/scripts/scripts.py]: maybe add a test for some of this stuff in https://github.com/ray-project/ray/blob/93ded5a3d533d4e2902ce17fe8547428b0966f6f/test/multi_node_test.py#L307

**Review by robertnishihara — CHANGES_REQUESTED** [review]

File: python/ray/worker.py [hunk 24]
```diff
@@ -1658,7 +1669,9 @@ def _init(address_info=None,
             plasma_directory=plasma_directory,
             huge_pages=huge_pages,
             include_webui=include_webui,
-            use_raylet=use_raylet)
+            use_raylet=use_raylet,
+            plasma_store_socket_name=plasma_store_socket_name,
+            temp_dir=tempfile_services.get_temp_root())
```
Comment (robertnishihara) [inline python/ray/worker.py]: in the `else` below, check that `temp_dir` is `None` and raise an exception if not

Comment (robertnishihara) [reply python/ray/worker.py]: same with `plasma_store_socket_name`

File: python/ray/worker.py [hunk 25]
```diff
@@ -2078,7 +2104,9 @@ def connect(info,
     assert worker.cached_remote_functions_and_actors is not None, error_message
     # Initialize some fields.
     worker.worker_id = random_string()
-
+    if temp_dir is not None:
+        # Override the temporary directory.
+        tempfile_services.set_temp_root(temp_dir)
```
Comment (robertnishihara) [inline python/ray/worker.py]: the only time we should call this is in `start_ray_processes`

Comment (robertnishihara) [reply python/ray/worker.py]: maybe default_worker.py also

**Comment by mitar** [comment]
@robertnishihara Should the socket path variable be "name" or "path"?

**Comment by robertnishihara** [comment]
@mitar either is ok I think. Path is probably a bit clearer, but we can change it later.

**Comment by suquark** [comment]
jenkins, retest this please

File: test/tempfile_test.py [hunk 26]
```diff
@@ -0,0 +1,43 @@
+import os
+import unittest
+import ray
+
+
+class TempFileTest(unittest.TestCase):
```
Comment (robertnishihara) [inline test/tempfile_test.py]: Sorry about this, but all new tests should use `pytest`. We're almost done converting all of the old tests to pytest.

Also, I'm not sure this warrants a whole new test file.

File: test/tempfile_test.py [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+        else:
+            self.fail("This test should raise an exception.")
+
+    def test_tempdir(self):
+        ray.init(temp_dir='/tmp/i_am_a_temp_dir')
+        self.assertTrue(os.path.exists('/tmp/i_am_a_temp_dir'),
+                        "Specified temp dir not found.")
+        ray.shutdown()
+        try:
+            os.rmdir('/tmp/i_am_a_temp_dir')
```
Comment (robertnishihara) [inline test/tempfile_test.py]: Also, please use `"` instead of `'`

Comment (robertnishihara) [reply test/tempfile_test.py]: This applies in other places in this file (and probably PR) as well

File: test/tempfile_test.py (hunk 26 above)
Comment (robertnishihara) [reply test/tempfile_test.py]: I think it's ok to have it as a separate file. I'm not sure what the best way to organize the tests is.

**Comment by raulchen** [comment]
I tried this on my laptop, but didn't find raylet log file under /tmp/ray/session_xxx/.

**Comment by suquark** [comment]
@raulchen What's your environment? I should test it.

**Comment by raulchen** [comment]
@suquark thanks, I thought I'd already set `redirect_worker_output=True`. Let me try again. 

Sorry for late response. My bandwidth is pretty tight recently. I roughly looked thru this PR, and it looked good overall. I'll take a closer look again tonight, and will merge it if no more problems. 

BTW, I think it's useful to also make raylet socket name configurable. Could you add that? thanks.

File: python/ray/services.py [hunk 28]
```diff
@@ -28,6 +24,14 @@
 import ray.local_scheduler
 import ray.plasma
 
+from ray.tempfile_services import (
+    set_temp_root, get_temp_root, get_logs_dir_path, get_ipython_notebook_path,
+    get_raylet_socket_name, get_temp_redis_config_path, new_redis_log_file,
+    new_raylet_log_file, new_local_scheduler_log_file, new_webui_log_file,
+    new_worker_log_file, new_log_monitor_log_file,
+    new_global_scheduler_log_file, new_plasma_store_log_file,
+    new_plasma_manager_log_file, new_monitor_log_file)
```
Comment (raulchen) [inline python/ray/services.py]: Nit, I prefer putting each of these items in one line, and sort them alphabetically. that would be more readable and maintainable.

File: python/ray/tempfile_services.py [hunk 29]
```diff
@@ -0,0 +1,290 @@
...
+import tempfile
+
+import ray.utils
+
+logger = logging.getLogger(__name__)
+_incremental_dict = collections.defaultdict(lambda: 0)
+_temp_root = None
+
+
+def make_inc_temp(suffix="", prefix="", directory_name="/tmp"):
```
Comment (raulchen) [inline python/ray/tempfile_services.py]: change default value of `directory_name` to `/tmp/ray`

File: python/ray/worker.py (hunk 25 above)
Comment (raulchen) [reply python/ray/worker.py]: I agree. @suquark could you fix this?

File: test/stress_tests.py [hunk 30]
```diff
@@ -164,10 +165,12 @@ def ray_start_reconstruction(request):
     plasma_addresses = []
     objstore_memory = plasma_store_memory // num_local_schedulers
     for i in range(num_local_schedulers):
-        store_stdout_file, store_stderr_file = ray.services.new_log_files(
-            "plasma_store_{}".format(i), True)
-        manager_stdout_file, manager_stderr_file = (ray.services.new_log_files(
-            "plasma_manager_{}".format(i), True))
+        store_stdout_file, store_stderr_file = (
+            ray.tempfile_services.new_log_files("plasma_store_{}".format(i),
```
Comment (raulchen) [inline test/stress_tests.py]: should this be `new_plasma_store_log_file`?

File: test/tempfile_test.py [hunk 31]
```diff
@@ -0,0 +1,95 @@
...
+import time
+import pytest
+import ray
+import ray.tempfile_services as tempfile_services
+
+
+def test_conn_cluster():
+    try:
+        ray.init(
+            redis_address="127.0.0.1:6379", temp_dir="/tmp/this_should_fail")
```
Comment (raulchen) [inline test/tempfile_test.py]: use this way to assert exceptions: https://stackoverflow.com/questions/23337471/how-to-properly-assert-that-an-exception-gets-raised-in-pytest

**Review by raulchen — COMMENTED** [review]
I left a few small comments, otherwise looks good to me.

**Review by raulchen — CHANGES_REQUESTED** [review]
there're some lint problems

**Comment by suquark** [comment]
Sorry for my late response. I happen to be quite busy today. Thanks again for your reviews and great patience!

File: python/ray/scripts/scripts.py [hunk 32]
```diff
@@ -177,11 +177,24 @@ def cli(logging_level, logging_format):
     is_flag=True,
     default=False,
     help="do not redirect non-worker stdout and stderr to files")
+@click.option(
+    "--plasma-store-socket-name",
+    default=None,
+    help="manually specify the socket name of the plasma store")
+@click.option(
+    "--raylet-socket-path",
```
Comment (raulchen) [inline python/ray/scripts/scripts.py]: This needs to be changed as well.

Comment (suquark) [reply python/ray/scripts/scripts.py]: I have amended the commit.

**Comment by raulchen** [comment]
Jenkins, retest this please

**Review by raulchen — APPROVED** [review]
LGTM. will merge when CI passes.

**Comment by suquark** [comment]
jenkins, retest this please.

File: test/tempfile_test.py [hunk 33]
```diff
@@ -0,0 +1,119 @@
...
+    assert os.path.exists(
+        "/tmp/i_am_a_temp_socket"), "Specified socket path not found."
+    ray.shutdown()
+    try:
+        os.remove("/tmp/i_am_a_temp_socket")
+    except Exception:
+        pass
+
+
+def test_raylet_tempfiles():
```
Comment (robertnishihara) [inline test/tempfile_test.py]: can you change all of the `'` to `"` in this function (and wherever else it is relevant?

**Comment by suquark** [comment]
This is pending for https://github.com/ray-project/ray/issues/3007.

File: python/ray/tempfile_services.py [hunk 34]
```diff
@@ -28,7 +29,13 @@ def make_inc_temp(suffix="", prefix="", directory_name="/tmp/ray"):
         "{directory_name}/{prefix}.{unique_index}{suffix}"
     """
     index = _incremental_dict[suffix, prefix, directory_name]
-    for seq in range(tempfile.TMP_MAX):
+    # `tempfile.TMP_MAX` could be extremely large,
+    # so we should use xrange in Python2.x
+    if sys.version_info[0] < 3:
```
Comment (robertnishihara) [inline python/ray/tempfile_services.py]: Instead of using `range` and `xrange`, can we just do `while index < tempfile.TMP_MAX`?

File: python/ray/tempfile_services.py [hunk 35]
```diff
@@ -0,0 +1,297 @@
...
+    index = _incremental_dict[suffix, prefix, directory_name]
+    # `tempfile.TMP_MAX` could be extremely large,
+    # so we should use xrange in Python2.x
+    if sys.version_info[0] < 3:
+        _range = xrange  # noqa: F821
+    else:
+        _range = range
+    for _ in _range(tempfile.TMP_MAX):
+        if index == 0:
+            file = os.path.join(directory_name, prefix + suffix)
```
Comment (robertnishihara) [inline python/ray/tempfile_services.py]: instead of using the variable name `file`, let's use something like `filename`

**Comment by suquark** [comment]
@robertnishihara It's ready to merge.

**Review by robertnishihara — APPROVED** [review]
I had to modify `.travis.yml` to test this in the xray test suite as well. I'll merge when the tests pass (in the morning).
